src/interceptors/ClientRequest/index.ts

```diff
@@ -1,61 +1,218 @@
+        const [url, options, callback] = normalizeClientRequestArgs(
+          'https:',
+          args
+        )
+
+        const mockAgent = new MockHttpsAgent({
+          customAgent: options.agent,
+          onRequest,
+          onResponse,
+        })
+        options.agent = mockAgent
+
+        return Reflect.apply(target, thisArg, [url, options, callback])
+      },
+    })
+
+    this.subscriptions.push(() => {
+      http.get = originalGet
+      http.request = originalRequest
+
+      https.get = originalHttpsGet
+      https.request = originalHttpsRequest
+    })
+  }
 
-        logger.info('native "%s" module restored!', protocol)
+  private onRequest: MockHttpSocketRequestCallback = async ({
+    request,
+    socket,
+  }) => {
+    const requestId = Reflect.get(request, kRequestId)
+    const { interactiveRequest, requestController } =
+      toInteractiveRequest(request)
+
+    // TODO: Abstract this bit. We are using it everywhere.
+    this.emitter.once('request', ({ requestId: pendingRequestId }) => {
+      if (pendingRequestId !== requestId) {
+        return
+      }
+
+      if (requestController.responsePromise.state === 'pending') {
+        this.logger.info(
+          'request has not been handled in listeners, executing fail-safe listener...'
+        )
+
+        requestController.responsePromise.resolve(undefined)
+      }
+    })
+
+    const listenerResult = await until(async () => {
+      await emitAsync(this.emitter, 'request', {
+        requestId,
+        request: interactiveRequest,
       })
 
-      const options: NodeClientOptions = {
-        emitter: this.emitter,
-        logger: this.logger,
+      return await requestController.responsePromise
+    })
+
+    if (listenerResult.error) {
+      // Treat thrown Responses as mocked responses.
+      if (listenerResult.error instanceof Response) {
+        socket.respondWith(listenerResult.error)
+        return
+      }
+
+      // Allow mocking Node-like errors.
+      if (isNodeLikeError(listenerResult.error)) {
+        socket.errorWith(listenerResult.error)
+        return
+      }
+
+      // Emit the "unhandledException" event to allow the client
+      //  to opt-out from the default handling of exceptions
+      // as 500 error responses.
+      if (this.emitter.listenerCount('unhandledException') > 0) {
+        await emitAsync(this.emitter, 'unhandledException', {
+          error: listenerResult.error,
+          request,
+          requestId,
+          controller: {
+            respondWith: socket.respondWith.bind(socket),
+            errorWith: socket.errorWith.bind(socket),
+          },
+        })
+
+        // After the listeners are done, if the socket is
+        // not connecting anymore, the response was mocked.
+        // If the socket has been destroyed, the error was mocked.
+        // Treat both as the result of the listener's call.
+        if (!socket.connecting || socket.destroyed) {
```

Comment:
So in the Socket world, the socket won't be `writableEnded` until `.end()` is called, which can take time (e.g. request body is streaming). 

Instead, I rely on `.connecting`, which we immediately set to `false` now in `.mockConnect()`, or the `.destroyed`, which will also immediately be set to false it you mock an error.